server/src/server.rs

```diff
@@ -461,91 +478,123 @@ impl GeyserService {
             SubscriptionClosedEvent::SlotUpdateSubscription(subscription_id) => {
                 let _ = slot_update_subscriptions.remove(&subscription_id);
             }
+            SubscriptionClosedEvent::ProgramUpdateSubscription(subscription_id) => {
+                let _ = program_update_subscriptions.remove(&subscription_id);
+            }
         }
 
         Ok(())
     }
 
     /// Streams account updates to subscribers.
     fn handle_account_update_event(
-        maybe_account_update: Result<AccountUpdate, crossbeam_channel::RecvError>,
+        maybe_account_update: Result<AccountUpdate, RecvError>,
         account_update_subscriptions: &HashMap<Uuid, AccountUpdateSubscription>,
         partial_account_update_subscriptions: &HashMap<Uuid, PartialAccountUpdateSubscription>,
+        program_update_subscriptions: &HashMap<Uuid, AccountUpdateSubscription>,
     ) -> GeyserServiceResult<Vec<Uuid>> {
         let account_update = maybe_account_update?;
 
-        let mut failed_subscription_ids = vec![];
-
-        let messages = (0..account_update_subscriptions.len())
-            .map(|_| account_update.clone())
-            .collect::<Vec<_>>();
-        failed_subscription_ids.extend(Self::send_account_update_messages(
-            account_update_subscriptions,
-            messages,
-        ));
-
-        let messages = (0..partial_account_update_subscriptions.len())
-            .map(|_| PartialAccountUpdate {
-                slot: account_update.slot,
-                pubkey: account_update.pubkey.clone(),
-                owner: account_update.owner.clone(),
-                is_startup: account_update.is_startup,
-                seq: account_update.seq,
-                tx_signature: account_update.tx_signature.clone(),
-                replica_version: account_update.replica_version,
+        let failed_account_update_sends: Vec<Uuid> = account_update_subscriptions
+            .iter()
+            .filter_map(|(uuid, sub)| {
+                if sub.accounts.contains(account_update.pubkey.as_slice())
+                    && matches!(
+                        sub.notification_sender.try_send(Ok(MaybeAccountUpdate {
```

Comment:
Wdyt about using `send`, return the future and do the same for the below subscriptions then `join_all` futures?

Reply:
wfm

Reply:
actually gotta create a runtime and do all that. its prob pretty fast as is, but will add a timestamp to these messages so we can log latencies